lib/DxilPIXPasses/DxilShaderAccessTracking.cpp

```diff
@@ -302,110 +337,401 @@ void DxilShaderAccessTracking::EmitAccess(LLVMContext &Ctx, OP *HlslOP,
+
+    SmallVector<llvm::Type*, 1> Elements{ Type::getInt32Ty(Ctx) };
+    llvm::StructType* UAVStructTy =
+        llvm::StructType::create(Elements, "class.RWStructuredBuffer");
+    std::unique_ptr<DxilResource> pUAV =
+        llvm::make_unique<DxilResource>();
+    pUAV->SetGlobalName((std::string("PIX_CountUAVName")+ std::to_string(UAVResourceHandle)).c_str());
+    pUAV->SetGlobalSymbol(UndefValue::get(UAVStructTy->getPointerTo()));
+    pUAV->SetID(UAVResourceHandle);
+    pUAV->SetRW(true); //sets UAV class
+    pUAV->SetSpaceID((
+        unsigned int)-2); // This is the reserved-for-tools register space
+    pUAV->SetSampleCount(1);
+    pUAV->SetGloballyCoherent(false);
+    pUAV->SetHasCounter(false);
+    pUAV->SetCompType(CompType::getI32());
+    pUAV->SetLowerBound(0);
+    pUAV->SetRangeSize(1);
+    pUAV->SetKind(DXIL::ResourceKind::RawBuffer);
+
+    auto pAnnotation =
+        DM.GetTypeSystem().GetStructAnnotation(UAVStructTy);
+    if (pAnnotation == nullptr) {
+
+        pAnnotation = DM.GetTypeSystem().AddStructAnnotation(UAVStructTy);
+        pAnnotation->GetFieldAnnotation(0).SetCBufferOffset(0);
+        pAnnotation->GetFieldAnnotation(0).SetCompType(
+            hlsl::DXIL::ComponentType::I32);
+        pAnnotation->GetFieldAnnotation(0).SetFieldName("count");
+    }
+
+    // Create handle for the newly-added UAV
+    if (IsDynamicResourceShaderModel(DM)) {
+      OP *HlslOP = DM.GetOP();
+      Function *CreateHandleFromBindingOpFunc =
+          HlslOP->GetOpFunc(DXIL::OpCode::CreateHandleFromBinding, Type::getVoidTy(Ctx));
+      Constant *CreateHandleFromBindingOpcodeArg =
+          HlslOP->GetU32Const((unsigned)DXIL::OpCode::CreateHandleFromBinding);
+      DxilResourceBinding binding =
+          resource_helper::loadBindingFromResourceBase(pUAV.get());
+      Value *bindingV = resource_helper::getAsConstant(
+          binding, HlslOP->GetResourceBindingType(), *DM.GetShaderModel());
+
+      Value *registerIndex = HlslOP->GetU32Const(UAVResourceHandle);
+
+      Value *isUniformRes = HlslOP->GetI1Const(0);
+
+      Value *createHandleFromBindingArgs[] = { CreateHandleFromBindingOpcodeArg, bindingV, registerIndex,
+                                              isUniformRes};
+
+      auto * handle = Builder.CreateCall(
+          CreateHandleFromBindingOpFunc,
+          createHandleFromBindingArgs,
+          name);
+
+      Function *annotHandleFn = HlslOP->GetOpFunc(DXIL::OpCode::AnnotateHandle, Type::getVoidTy(Ctx));
+      Value *annotHandleArg = HlslOP->GetI32Const((unsigned)DXIL::OpCode::AnnotateHandle);
+      DxilResourceProperties RP = resource_helper::loadPropsFromResourceBase(pUAV.get());
+      Type *resPropertyTy = HlslOP->GetResourcePropertiesType();
+      Value *propertiesV = resource_helper::getAsConstant(RP, resPropertyTy, *DM.GetShaderModel());
+
+      unsigned int ID = DM.AddUAV(std::move(pUAV));
+      assert((unsigned)ID == UAVResourceHandle);
+      (void)ID;
```

Comment:
if you opt to use DXASSERTs, DXASSERT_LOCALVAR is made for situations like this

Reply:
Thanks. I honestly wasn't sure what the pattern was. I should have looked harder/asked.